patches/server/0902-OfflinePlayer-advancements.patch

```diff
@@ -0,0 +1,59 @@
+Date: Thu, 22 Apr 2021 01:41:56 -0700
+Subject: [PATCH] OfflinePlayer advancements
+
+
+diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
+index 188c68d4ae46bc05c6d9c901b5c6ab883fa05ea4..84707cb3a67a3000ebefc8e140209988f56b815f 100644
+--- a/src/main/java/net/minecraft/server/players/PlayerList.java
++++ b/src/main/java/net/minecraft/server/players/PlayerList.java
+@@ -1444,7 +1444,12 @@ public abstract class PlayerList {
+ 
+     public PlayerAdvancements getPlayerAdvancements(ServerPlayer player) {
+         UUID uuid = player.getUUID();
+-        PlayerAdvancements advancementdataplayer = (PlayerAdvancements) player.getAdvancements(); // CraftBukkit
++        // Paper start
++        return this.getPlayerAdvancements(uuid, player);
++    }
++    public PlayerAdvancements getPlayerAdvancements(UUID uuid, @Nullable ServerPlayer player) {
++        PlayerAdvancements advancementdataplayer = player == null ? null : player.getAdvancements(); // CraftBukkit
++        // Paper end
+ 
+         if (advancementdataplayer == null) {
+             File file = this.server.getWorldPath(LevelResource.PLAYER_ADVANCEMENTS_DIR).toFile();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index 1f2bc88d4570c6ef00e67a772b745e0b0c98e051..da495780e4f18f96341686b37daa514d39f2adc7 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -300,6 +300,30 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+ 
+         return result;
+     }
++
++    @Override
++    public org.bukkit.advancement.AdvancementProgress getAdvancementProgress(org.bukkit.advancement.Advancement advancement) {
++        Player player = this.getPlayer();
++        if (player != null) return player.getAdvancementProgress(advancement);
++        com.google.common.base.Preconditions.checkArgument(advancement != null, "advancement");
++
++        org.bukkit.craftbukkit.advancement.CraftAdvancement craft = (org.bukkit.craftbukkit.advancement.CraftAdvancement) advancement;
++        net.minecraft.server.PlayerAdvancements data = this.server.getHandle().getPlayerAdvancements(this.getUniqueId(), null);
++        net.minecraft.advancements.AdvancementProgress progress = data.getOrStartProgress(craft.getHandle());
++        return new org.bukkit.craftbukkit.advancement.CraftAdvancementProgress(craft, data, progress) {
++            @Override
++            public boolean awardCriteria(String criteria) {
++                throw new UnsupportedOperationException("Cannot award criteria to an offline player");
++            }
++
++            @Override
++            public boolean revokeCriteria(String criteria) {
++                boolean result = super.revokeCriteria(criteria);
```

Comment:
shouldn't this also be restricted?

Reply:
Well it should work fine with offline players. the issue with awarding is there's a whole bunch of other logic run if a criteria award complete some task.